File: tests/matlab/crc32.m
classdef crc32
    %UNTITLED Summary of this class goes here
    %   Detailed explanation goes here
    
    properties(Constant = true, Access = private)
        Lut@uint32 = crc32.GenerateLut(hex2dec('EDB88320'))
    end
    
    methods(Static = true, Access = private)
        function lut = GenerateLut(Polynomial)
            input = uint32([0:255]);
            output = input;
            for i = 1:8
                shift = bitshift(output, -1);
                output = bitxor(shift, bitand(output, uint32(1)) * Polynomial);
            end
            lut = output;
        end
    end
    
    methods
        function this = crc32(~)
        end
        function checksum = Calculate(this, checksum, data)
            assert(isa(checksum, 'uint32') && isscalar(checksum), 'Value must be ''uint32 scalar''.');
            assert(isa(data, 'uint8'), 'Value must be ''uint8 vector''.');
            lut = this.Lut;
            crc = bitcmp(checksum);
            for ii = 1:numel(data)
                crc = bitxor(crc, uint32(data(ii)));
                tmp1 = uint32(floor(double(crc) / (256)));
                tmp2 = lut(1 + bitand(crc, uint32(255)));
                crc = bitxor(tmp1, tmp2);
            end
            checksum = bitcmp(crc);
        end
    end
end

File: tests/python/crc32.py
# -*- coding: utf-8 -*-
"""
@author: msquirogac
"""

class crc32(object):
    __slots__ = ['lut']
    def __init__(self):
        self.lut = self.GenerateLut(0xEDB88320)

    @staticmethod
    def GenerateLut(polynomial):
        output = []
        for ii in range(256):
            elem = ii
            for j in range(8):
                shift = elem >> 1
                elem = shift ^ (elem & 1) * polynomial
            output.append(elem & 0xFFFFFFFF)
        return tuple(output);

    def Calculate(self, crc, data):
        length = len(data)
        crc = (crc ^ 0xFFFFFFFF)
        for ii in range(length):
            crc = crc ^ data[ii]
            crc = (crc >> 8) ^ (self.lut[crc & 0xFF])
        return (crc ^ 0xFFFFFFFF)
